R/run-nmtran.R

```diff
@@ -0,0 +1,223 @@
+#'  (if 0, don't pass `-maxlim` to nmfe) (default is 2)
+#'    - If `maxlim = 3`, it is preferred to also set `tprdefault = TRUE`, but
+#'    _not_ `prdefault`, as `NM-TRAN`'s optional resizing of the `PREDPP` size
+#'    parameter `MAXRECID` may conflict with the `-prdefault` option.
+#'  - **`prdefault`**: If `TRUE`, do not recompile any routines other than `FSUBS`
+#'  - **`tprdefault`**: If `TRUE`, test if is okay to do `-prdefault`
+#'
+#' @param .mod A `bbr` model object.
+#' @param .bbi_args A named list specifying arguments to pass to `NM-TRAN`.
+#'   Similar to the `.bbi_args` argument defined in [submit_model()], though here
+#'   only `prdefault`, `tprdefault`, and `maxlim` flags are passed to `NM-TRAN`.
+#'   See details.
+#' @param .config_path Path to a bbi configuration file. If `NULL`, the default,
+#'   will attempt to use a `bbi.yaml` in the same directory as the model.
+#' @param nmtran_exe Path to an `NM-TRAN` executable. If `NULL`, will look for a
+#'   `bbi.yaml` file in the same directory as the model.
+#' @param delete_on_exit Logical. If `FALSE`, don't delete the temporary folder
+#'   containing the `NM-TRAN` run, which will be stored in the current working
+#'   directory.
+#'
+#' @examples
+#' \dontrun{
+#' mod <- read_model(file.path(MODEL_DIR, 1))
+#' run_nmtran(mod)
+#'
+#' # Set the path to an NM-TRAN executable
+#' run_nmtran(mod, nmtran_exe = "/opt/NONMEM/nm75/tr/NMTRAN.exe")
+#'
+#' }
+#' @name nmtran
+NULL
+
+
+#' @describeIn nmtran Run `NM-TRAN` on a model object to validate its control
+#' stream for correct coding before submission.
+#' @export
+run_nmtran <- function(
+    .mod,
+    .bbi_args = list(prdefault = FALSE, tprdefault = FALSE, maxlim = 2),
+    .config_path = NULL,
+    nmtran_exe = NULL,
+    delete_on_exit = TRUE
+){
+  check_model_object(.mod, "bbi_nonmem_model")
+
+  # Capture NONMEM and NMTRAN options
+  nmtran_exe <- locate_nmtran(.mod, .config_path, nmtran_exe)
+  nm_ver <- attr(nmtran_exe, "nonmem_version")
+
+  # Combine NONMEM submission args
+  #  - The main ones of interest are prdefault, tprdefault, and maxlim, which
+  # impact the evaluation of `NM-TRAN`
+  .bbi_args <- parse_args_list(.bbi_args, .mod[[YAML_BBI_ARGS]])
+  .bbi_args <- Filter(Negate(is.null), .bbi_args[c("prdefault", "tprdefault", "maxlim")])
+  cmd_args <- check_bbi_args(.bbi_args)
+
+
+  mod_path <- get_model_path(.mod)
+  data_path <- get_data_path_from_ctl(.mod)
+
+  # make temporary directory in current directory
+  mod_name <- fs::path_ext_remove(basename(mod_path))
+  temp_folder <- paste0("nmtran_", mod_name, "_", basename(tempdir()))
+  if(fs::dir_exists(temp_folder)) fs::dir_delete(temp_folder)
+  fs::dir_create(temp_folder)
+  if(isTRUE(delete_on_exit)){
+    on.exit(unlink(temp_folder, recursive = TRUE, force = TRUE))
+  }
+
+  # Copy model
+  file.copy(mod_path, temp_folder, overwrite = TRUE)
+  nmtran_mod <- new_model(file.path(temp_folder, basename(mod_path)), .overwrite = TRUE)
```

Comment:
Hmm, I think this `new_model` call is just for the purposes of `modify_data_path_ctl`. That's okay as is, but it'd be easy to split out the main work from `modify_data_path_ctl` so that you had a function that took in a `nmrec_ctl_records` object and modified it. `run_nmtran` could then call that write the result wherever it wanted with `nmrec::write_ctl`.

Again, mostly just thinking aloud, though I do think that'd lead to nicer separation of the main functionality from unrelated model/disk io details.